### PR TITLE
Warnings for divergences and max treedepth

### LINKED ISSUE #41: sample should explicitly print warning messages when models have divergences or hit max_treedepth
Right now it doesn't.

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Fixes #41 

This PR is currently done to the read_csv branch so the differences are more clear. Need to switch the base PR branch once #95 is merged.

After sampling cmdstanr will now output something like:

```
18 of 1000 (2%) transitions ended with a divergence.
These divergent transitions indicate that HMC is not fully able to explore the posterior distribution.
Try increasing adapt delta closer to 1.
If this doesn't remove all divergences, try to reparameterize the model.

196 of 1000 (20%) transitions hit the maximum treedepth limit of 5 or 2^5 leapfrog steps.
Trajectories that are prematurely terminated due to this limit will result in slow exploration.
For optimal performance, increase this limit.
```

I need to add a few tests but otherwise, this is good for review.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
Rok Češnovar, Univ. of Ljubljana

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -367,3 +367,40 @@ set_num_threads <- function(num_threads) {
          call. = FALSE)
   }
 }
+
+check_divergences <- function(data_csv) {
+  divergences <- posterior::extract_one_variable_matrix(data_csv$sampler, "divergent__")
+  num_of_divergences <- sum(divergences)
+  if (num_of_divergences > 0) {
+    if(data_csv$sampling_info$sample_save_warmup == 1) {
```
Comment (bbbales2) [inline R/utils.R]: Don't report divergences during warmup. These happen a lot cause the timestep is adjusting (divergences will cause the timestep to go down).

File: R/utils.R [hunk 2]
```diff
@@ -367,3 +367,40 @@ set_num_threads <- function(num_threads) {
...
+    if (max_treedepth_hit > 0) {
+      if(data_csv$sampling_info$sample_save_warmup == 1) {
+      num_iter <- data_csv$sampling_info$sample_num_samples + data_csv$sampling_info$sample_num_warmup
+    } else {
+      num_iter <- data_csv$sampling_info$sample_num_samples
+    }
+    percentage_max_treedepth <- (max_treedepth_hit)/num_iter*100
+    message(max_treedepth_hit, " of ", num_iter, " (", (format(round(percentage_max_treedepth, 0), nsmall = 0)), "%)",
+            " transitions hit the maximum treedepth limit of ", max_treedepth,
+            " or 2^", max_treedepth, " leapfrog steps.\n",
```
Comment (bbbales2) [inline R/utils.R]: 2^max_treedepth - 1 (I think 1023 is the max number of steps for max_treedepth = 10 )

File: R/utils.R [hunk 3]
```diff
@@ -367,3 +367,40 @@ set_num_threads <- function(num_threads) {
...
+  }
+}
+
+check_sampler_transitions_treedepth <- function(data_csv) {
+  max_treedepth <- data_csv$sampling_info$sample_adapt_hmc_nuts_max_depth
+  treedepth <- posterior::extract_one_variable_matrix(data_csv$sampler, "treedepth__")
+  max_treedepth_hit <- sum(treedepth >= max_treedepth)
+    if (max_treedepth_hit > 0) {
+      if(data_csv$sampling_info$sample_save_warmup == 1) {
+      num_iter <- data_csv$sampling_info$sample_num_samples + data_csv$sampling_info$sample_num_warmup
```
Comment (bbbales2) [inline R/utils.R]: Same with treedepth warnings

File: R/utils.R [hunk 4]
```diff
@@ -367,3 +367,40 @@ set_num_threads <- function(num_threads) {
...
+    } else {
+      num_iter <- data_csv$sampling_info$sample_num_samples
+    }
+    percentage_max_treedepth <- (max_treedepth_hit)/num_iter*100
+    message(max_treedepth_hit, " of ", num_iter, " (", (format(round(percentage_max_treedepth, 0), nsmall = 0)), "%)",
+            " transitions hit the maximum treedepth limit of ", max_treedepth,
+            " or 2^", max_treedepth, " leapfrog steps.\n",
+            "Trajectories that are prematurely terminated due to this limit will result in slow exploration.\n",
+            "For optimal performance, increase this limit.")
```
Comment (bbbales2) [inline R/utils.R]: Maybe replace "For optimal performance, increase this limit" maybe say, "Increasing the max_treedepth limit can avoid this at the expense of more computation."

And then maybe add the same line as with divergences: "If increasing max_treedepth does not remove warnings, try to reparameterize the model.\n"

File: R/utils.R (hunk 2 above)
Comment (rok-cesnovar) [reply R/utils.R]: Ah, I was following what cmdstan diagnose outputs, but yeah, you are right.

File: R/utils.R (hunk 4 above)
Comment (rok-cesnovar) [reply R/utils.R]: Thanks. I added this. 

I used max_depth as that is the name of the argument for now.

File: R/utils.R [hunk 5]
```diff
@@ -387,20 +390,24 @@ check_divergences <- function(data_csv) {
...
+  if (max_treedepth_hit > 0) {
+    num_iter <- data_csv$sampling_info$sample_num_samples
     percentage_max_treedepth <- (max_treedepth_hit)/num_iter*100
     message(max_treedepth_hit, " of ", num_iter, " (", (format(round(percentage_max_treedepth, 0), nsmall = 0)), "%)",
             " transitions hit the maximum treedepth limit of ", max_treedepth,
-            " or 2^", max_treedepth, " leapfrog steps.\n",
+            " or 2^", max_treedepth, "-1 leapfrog steps.\n",
             "Trajectories that are prematurely terminated due to this limit will result in slow exploration.\n",
-            "For optimal performance, increase this limit.")
+            "Increasing the max_treedepth limit can avoid this at the expense of more computation.\n",
```
Comment (bbbales2) [inline R/utils.R]: Change "Increasing the max_treedepth limit" to "Increasing max_depth"

**Comment by rok-cesnovar** [comment]
Yeah. I already use CSV files for the read_csv branch (This PR builds of the `read_csv` branch in PR #95), will just read those. I think they should have divergences as well as max treedepth hits. If not I will just add them manually.

**Comment by bbbales2** [comment]
I see tests. Want me to review?

**Comment by rok-cesnovar** [comment]
One of the failed-chains tests is failing for some yet unknown reason. I am going to try and fix that and will tag you then. Hopefully should be soon. Thanks @bbbales2!

**Comment by rok-cesnovar** [comment]
Should be ready now.

**Comment by bbbales2** [comment]
I added a check to make sure divergences/max treedepth warnings in warmup don't get counted.

In the process I needed to change a csv check and I think there was a problem with test-csv.R. Check it out and see what you think.

The file here needs updated if indeed you agree the changes are correct:

```
sampler_params <- readRDS(test_path("answers", "rstan-read-stan-csv-sampler-params.rds"))
```

(I wasn't sure how to do it/was too lazy)

**Comment by rok-cesnovar** [comment]
Thanks, I updated the rds, this should pass now. Should we merge this to the read_csv if it does?

**Comment by bbbales2** [comment]
If you're happy I'm happy. If the tests pass merge away.

**Pull request merged**